docs/architecture.md

```diff
@@ -0,0 +1,223 @@
+
+```
+void world_update(world) {
+    gravity_system(world);
+    move_system(world);
+}
+
+void gravity_system(world) {
+    for (each entity with mass and velocity) {
+        // accelerate the masses together
+    }
+}
+
+void move_system(world) {
+    for (each entity with velocity and position) {
+        position += velocity
+    }
+}
+
+...
+```
+
+Systems are called in a specific order to update the entire world to the next frame. Some systems may depend on each other, and other systems can run in parallel.
+
+The EnTT library provides many features making iteration convenient.
+
+### Responsibility of Systems
+
+Ideally, components are raw data that don't describe any behaviour. However, it's acceptable for them to have constructors, destructors, member functions, and RAII members as long as components are only modified by very specific systems. This includes creating and destroying components. Many systems reading the same component is okay.
+
+EnTT allows polymorphically accessing components, to allow for deleting, copying, and listing all components an entity has. These features should never be used outside of debugging, since they potentially modify components and calls functions outside of their respective systems.
+
+In short, **components must only be modified by a few select systems**. There's a few other practical reasons why:
+
+* Components may need to be synced with external references: pointers, GPU resourses, or physics engine objects. Some of these require specific threads.
+* Multiple Systems that modify different components can run in parallel without mutexes.
+
+There's a few important operations that need to be addressed however:
+
+* **Deleting entire entities**: Deleting an entity means modifying every component container the entity is associated with. Entities can be marked for deletion with a component, or external queue. 'Deleter' Systems can then delete their respective components near the end of the frame.
+* **Copying entire entities**: It may seem convenient to iterate all of an entity's components and copy them into another entity, but this breaks the rules. Copies of an entity should be created by the same system as the original entity. A dedicated 'mark for copy' and copy systems may be practical depending on the case.
+
+### A solution to all world state?
+
+Entities and Components are powerful, but cannot solve every single storage use case alone. EnTT specifically, only allows one component of a unique type per entity, and only one container of these components (pool) per world (registry). One case may require a variable number of the same component for an entity, or the problem may not be well suited for entities and components at all.
+
+EnTT features customizability and future versions can be more versitile, but either way, ECS is best when combined with other data structures besides just components. For example, queues can be used to pass messages between systems.
```

Comment:
Maybe also mention lockless multithreaded queues to support systems running on multiple threads.